### PR TITLE
Java: Cast to byte or short after +, -

### PR DESCRIPTION
**Pull request by lukemaurer**
Test included.

### CODE REVIEW

File: Source/Dafny/Compiler-java.cs [hunk 1]
```diff
@@ -2087,6 +2087,16 @@ protected override void EmitActualTypeArgs(List<Type> typeArgs, Bpl.IToken tok,
             opString = "+";
           } else if (resultType is UserDefinedType && !resultType.IsIntegerType && !unsignedTypes.Contains(TypeName(e0.Type, null, tok)) && !unsignedTypes.Contains(TypeName(e1.Type, null, tok))) {
             opString = "+";
+            switch (TypeName(resultType, null, tok)) {
```
Comment (seanmcl) [inline Source/Dafny/Compiler-java.cs]: Could you add a reference to somewhere explaining why we're casting here.  E.g. https://docs.oracle.com/javase/specs/jls/se7/html/jls-5.html#jls-5.6.2

File: Test/comp/NativeNumbers.dfy [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+method Main() {
+    Int8Test();
+    Int16Test();
+}
+
+// test handling of byte/short arithmetic in Java
+newtype {:nativeType "sbyte"} int8 = x | -0x80 <= x < 0x80
+newtype {:nativeType "short"} int16 = x | -0x8000 <= x < 0x8000
+
+method Int8Test() {
```
Comment (seanmcl) [inline Test/comp/NativeNumbers.dfy]: Could we also test the boundaries here, in particular specify what should happen when a/b are ~2^8?

Comment (lukemaurer) [reply Test/comp/NativeNumbers.dfy]: Well, these are signed integers; is it still valuable to pass in, say, 126 and 127?

File: Test/comp/NativeNumbers.dfy.expect [hunk 3]
```diff
@@ -0,0 +1,18 @@
+
+Dafny program verifier finished with 7 verified, 0 errors
```
Comment (seanmcl) [inline Test/comp/NativeNumbers.dfy.expect]: Why are there 3 blocks here?  I see only two tests, one for 8 and 16 bits resp.

File: Test/comp/NativeNumbers.dfy (hunk 2 above)
Comment (seanmcl) [reply Test/comp/NativeNumbers.dfy]: Not sure what's the most valuable.  I was just thinking I'd like to see the number wrap around, e.g. adding 1 to 2^7-1 should yield -2^7.  Or maybe Dafny doesn't allow wrapping like that because the verifier will give an error and it won't make it to the compiler?  (Forgive my lack of Dafny knowledge.)  I just don't yet see how this test is testing the type cast in Java.

Comment (lukemaurer) [reply Test/comp/NativeNumbers.dfy]: Oh—right, only casts that are known to be safe (that is, in bounds) will make it past the verifier.  These tests are just meant to make sure that code generation is keeping everything straight: declaring the right types, using the right operators, etc.

Comment (RustanLeino) [reply Test/comp/NativeNumbers.dfy]: Integer types never wrap in Dafny, and any `as` type cast is allowed only if the source expression already has a value in the target type.

Bitvector types in Dafny also have arithmetic operations, and they wrap. All bitvector values are unsigned. If this is useful for the bug that was fixed (or just as additional tests), then you may consider tests along the lines of

    var a: bv8 := 250;
    a := a + 6;
    assert a == 0;

    var b: bv7 := 126;
    b := b + 5;
    assert b == 3;

    print a, " ", b, "\n";

  Rustan

Comment (seanmcl) [reply Test/comp/NativeNumbers.dfy]: Are bitvectors represented in Java as small types like `short` if the vector is small enough, or do we make a Java array?  I thought it was the later, but If the former, that would be a cool test.  
@lukemaurer I'm fine with this test as it is now if we can't test with bits.

Comment (lukemaurer) [reply Test/comp/NativeNumbers.dfy]: Small bitvectors get represented as primitives wrapped in our special unsigned wrapper class, which means they aren't affected by this patch.  Could be useful to have the tests for if that changes, though.

Comment (RustanLeino) [reply Test/comp/NativeNumbers.dfy]: The C# version uses unsigned integers of built-in lengths, if possible, and uses something bigger (`BigInteger`, I seem to recall) for bitvectors of more than 64 bits. From what Luke is saying, it sounds like the Java version attempts to do something similar, but uses the special unsigned wrapper classes. If that needs to change, let's do that as a separate PR (in due time).

File: Source/Dafny/Compiler-java.cs [hunk 4]
```diff
@@ -2087,6 +2087,22 @@ protected override void EmitActualTypeArgs(List<Type> typeArgs, Bpl.IToken tok,
             opString = "+";
           } else if (resultType is UserDefinedType && !resultType.IsIntegerType && !unsignedTypes.Contains(TypeName(e0.Type, null, tok)) && !unsignedTypes.Contains(TypeName(e1.Type, null, tok))) {
             opString = "+";
+            // If operating on bytes or shorts, the operation will widen both
+            // to ints and the result will be an int, so we need to cast back.
```
Comment (seanmcl) [inline Source/Dafny/Compiler-java.cs]: Nice comment, thanks.
